src/modules/haskell.rs

```diff
@@ -0,0 +1,149 @@
+    let mut module = context.new_module("haskell");
+    let config = HaskellConfig::try_load(module.config);
+
+    let is_hs_project = context
+        .try_begin_scan()?
+        .set_files(&config.detect_files)
+        .set_extensions(&config.detect_extensions)
+        .set_folders(&config.detect_folders)
+        .is_match();
+
+    if !is_hs_project {
+        return None;
+    }
+
+    let parsed = StringFormatter::new(config.format).and_then(|formatter| {
+        formatter
+            .map_meta(|var, _| match var {
+                "symbol" => Some(config.symbol),
+                _ => None,
+            })
+            .map_style(|variable| match variable {
+                "style" => Some(Ok(config.style)),
+                _ => None,
+            })
+            .map(|variable| match variable {
+                "version" => get_version(context).map(Ok),
+                "ghc_version" => get_ghc_version(context).map(Ok),
+                "snapshot" => get_snapshot(context).map(Ok),
+                _ => None,
+            })
+            .parse(None, Some(context))
+    });
+
+    module.set_segments(match parsed {
+        Ok(segments) => segments,
+        Err(error) => {
+            log::warn!("Error in module `haskell`:\n{}", error);
+            return None;
+        }
+    });
+
+    Some(module)
+}
+
+fn get_ghc_version(context: &Context) -> Option<String> {
+    Some(
+        context
+            .exec_cmd("ghc", &["--numeric-version"])?
+            .stdout
+            .trim()
+            .to_string(),
+    )
+}
+
+fn get_snapshot(context: &Context) -> Option<String> {
+    if !is_stack_project(context) {
+        return None;
+    }
+    let file_contents = utils::read_file(context.current_dir.join("stack.yaml")).ok()?;
+    let yaml = yaml_rust::YamlLoader::load_from_str(&file_contents).ok()?;
+    let version = yaml.first()?["resolver"]
+        .as_str()
+        .or_else(|| yaml.first()?["snapshot"].as_str())
+        .filter(|s| s.starts_with("lts") || s.starts_with("nightly") || s.starts_with("ghc"))
+        .unwrap_or("<custom snapshot>");
+    Some(version.to_string())
+}
+
+fn get_version(context: &Context) -> Option<String> {
+    get_snapshot(context).or_else(|| get_ghc_version(context))
+}
+
+fn is_stack_project(context: &Context) -> bool {
+    match context.dir_contents() {
+        Ok(dir) => dir.has_file_name("stack.yaml"),
+        Err(_) => false,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::test::ModuleRenderer;
+    use ansi_term::Color;
+    use std::fs::File;
+    use std::io;
+    use std::io::Write;
+
+    #[test]
+    fn folder_without_hs_files() -> io::Result<()> {
+        let dir = tempfile::tempdir()?;
+        let actual = ModuleRenderer::new("haskell").path(dir.path()).collect();
+        let expected = None;
+        assert_eq!(expected, actual);
+        dir.close()
+    }
+
+    #[test]
+    fn folder_stack() -> io::Result<()> {
```

Comment:
Can we also add a test case here with a more polluted YAML content? Something that we might see in real life, e.g. with packages and flags. Could be a toy project with a few dependencies, or just something pulled from a random repo.

I'm not so concerned about it for this specific PR, but it would be a small amount of protection against future breakage.

Reply:
Copy-pasted a stack file from Reanimate. https://github.com/reanimate/reanimate/blob/master/stack-lts-14.yaml